### ISSUE TITLE
IOBE: no procedure provided to prepare the first (configuration) log entry

### ISSUE THREAD

**Issue by amiloradovsky** [canonical/raft#82]
To implement a custom I/O backend (`struct raft_io`), at the `bootstrap` phase, one needs to encode the configuration into an internal (for libraft) format, to write it into the persistent storage as the first record. But currently there aren't exposed any procedures for that.

It would suffice to just add this into `include/raft.h`:
```C
RAFT_API int
configurationEncode (const struct raft_configuration *c,
                     struct raft_buffer *buf);
```
then use it with
```C
  struct raft_entry entry;
  entry.term = 1;
  entry.type = RAFT_CHANGE;
  entry.buf = buffer;  /* encoded configuration */
  entry.batch = NULL;  /* OK? */
```
and store the entry via the usual means provided by the new IOBE.
Although there may be a better solutions, or I just miss something.

**Comment by freeekanayaka** [comment]
@amiloradovsky you're right.

The reason I didn't yet add encode/decode functions to the public API is that I'm not entirely sure if the encoding should instead be customizable.

However, since the public API of this raft library has not reached yet a 1.0 stable version, I think it's fine expose the function you need in order to get unblocked, with the understanding that interfaces might change between now and the final 1.0 version.

I think I asked you already, but I didn't get a replay: would you mind to elaborate why you need to implement a custom ```raft_io``` backend? That's one of the hardest parts of the library and a lot of work to get done right. I'd like to understand what use cases are currently not met by the stock backend based on ```libuv```.

**Comment by amiloradovsky** [comment]
@freeekanayaka Ah, yes, sorry for the silence. There are a lot of reasons, small and big:

- The libuv-based IOBE seems flakier than the rest of the library: GCC and Valgrind warnings lead to there; when the application crashes, it seems to be due to an unitialized values in there (hard to follow).
- Also, I heed to use (C)ZMQ in the other parts of the system, so it would be nice to be able to just drop libuv as a (relatively big) dependency.
- The libuv IOBE persistent storage needs to be cleared (`rm <dir>/*`) before each run, or it won't bootstrap.
- The burden of aligning the message size is put on the user (me) and not handled automatically somewhere under the hood.
- It doesn't support IPv6, or any other protocols than TCP/IPv4.
- I'm not interested in either Windows or the mobile OSes, and would prefer less complexity.
- Also, libuv is distributed under a soup of BSD-like licenses, I'd prefer something more standard, like ASL or MPL.
- etc. etc.

Presently, I still use the provided libuv IOBE, because my own implementation needs (a lot of?) debugging (for instance, it can't elect a leader, despite all the messages seem to be sent and received, and the ticks being on time) and further polishing (use CZMQ frames instead of manual pointer shifts), and this isn't a priority right now. Nevertheless, I'm planning to finish it later, to get a smaller and more closed system. Or be able to compare my own implementation against another (for testing etc.).

As for the practical application, it's for an in-cluster synchronization for a storage system.

**Comment by freeekanayaka** [comment]
> @freeekanayaka Ah, yes, sorry for the silence.

No worries! Thanks for following up.

>     * The libuv-based IOBE seems flakier than the rest of the library: GCC and Valgrind warnings lead to there; when the application crashes, it seems to be due to an unitialized values in there (hard to follow).

Yeah, I don't think there are any memory leaks or actual memory issues **in practice**, since the ```libuv``` IOBE is fairly well unit tested and was used extensively in production LXD systems for quite a bit now.

That being said, I agree that IOBE needs more work to get up the same quality level as the core system. One reason is partly how complicated it turns to be to actually manage asynchronous I/O, especially in C. Still learning myself, expect improvements  in the ```libuv``` backend in the coming weeks and months.

> 
>     * Also, I heed to use (C)ZMQ in the other parts of the system, so it would be nice to be able to just drop libuv as a (relatively big) dependency.
> 

That's understandable.

>     * The libuv IOBE persistent storage needs to be cleared (`rm <dir>/*`) before each run, or it won't bootstrap.

I don't understand this. You surely don't want to bootstrap a single raft instance more than once? So why do you need to `rm <dir>/*`? I'm confused.

>     * The burden of aligning the message size is put on the user (me) and not handled automatically somewhere under the hood.

I'm not sure to understand this either, mind elaborating?

>     * It doesn't support IPv6, or any other protocols than TCP/IPv4.
> 

Yeah, my wish would be at least implement TLS support.

>     * I'm not interested in either Windows or the mobile OSes, and would prefer less complexity.
> 

Sounds good too. Although ```libuv``` on Linux is really a thin and neat wrapper over ```epoll``` and doesn't add much complexity imo.

>     * Also, libuv is distributed under a soup of BSD-like licenses, I'd prefer something more standard, like ASL or MPL.
> 

This really depends on the use you have to make of it. BSD is a fairly liberal license, so I'd be surprised if it creates any issue for your project. However I understand wanting clarity of licenses for your dependencies.

> 
> As for the practical application, it's for an in-cluster synchronization for a storage system.

That sounds very cool. I'm happy to support you in whatever you need, as I'd really like to see another consumer of this raft library, beyond dqlite.

One thing I'd would ideally like to do if I had time would be to breakdown the ```libuv``` IOBE into separate re-usable components and abstractions. So for example when one wants to implement a new backend, only things like low-level transport mechanisms and disk persistence would need to be implemented. But I think that's quite a bit of work to do right.

Probably my suggestion would be to stick to the ```libuv``` IOBE for now, if you can, then when its quality and internal code becomes more mature over the next weeks or months you might take a chance to basically copy-paste it and replace all ```libuv``` specific things with whatever you want. That might probably give us a better clue what parts need to be abstracted in order to make it easier to implement new backends.

**Comment by amiloradovsky** [comment]
@freeekanayaka As for removing the persistent data, that's simply the solution I've found to make `bootstrap` happy. Maybe I should just check first, and bootstrap only if the directory is empty.

As for the alignment, I just can't add an entry which size isn't multiple of 8. Sure that's justifiable by the memory allocation & access issues. But that should be handled elsewhere.
Currently, I have to place the actual length of the message in the first few octets, then the message itself (unaligned), then zero the rest, and only then can submit to apply it.
I guess these problems should be resolved implicitly, behind the public API. So one could just call `raft_apply` with a buffer of any length, not necessarily multiple of something, and it would be only advisory to have the buffer size already aligned (no need for another `*alloc` and `mem(p)cpy`).

I have a few minor fixes for the warnings. Just to be sure everything's okay there, and nothing really important is drowning in the flood. Plus, e.g. Nixpkgs have `-Werror` enabled by default.

And also it may be a good idea to bump the version number a little. So I could update the Guix port (they don't like when packages are versioned by commit ids).

And finally, I need a liveness monitoring mechanism implemented somewhere, and thought  it could be fitted relatively easily into libraft itself:
Not only the leader has to send the heartbeats to the followers, but also the followers have to send a heartbeat to the leader, or it will add an entry saying the silent node is out (and that in turn will lead to a call back, provided by the FSM). But it's likely the topic for another "issue" (feature request).
